// @clerk/react-router/dist/server/clerkMiddleware.js
// src/server/clerkMiddleware.ts
import { AuthStatus, constants, createClerkRequest } from "@clerk/backend/internal";
import { handleNetlifyCacheInDevInstance } from "@clerk/shared/netlifyCacheHandler";
import { createContext } from "react-router";
import { clerkClient } from "./clerkClient.js";
import { loadOptions } from "./loadOptions.js";
import { patchRequest } from "./utils.js";
var authFnContext = createContext(null);
var requestStateContext = createContext(null);
var clerkMiddleware = (options) => {
  return async (args, next) => {
    const clerkRequest = createClerkRequest(patchRequest(args.request));
    const loadedOptions = loadOptions(args, options);
    const {
      apiUrl,
      secretKey,
      jwtKey,
      proxyUrl,
      isSatellite,
      domain,
      publishableKey,
      machineSecretKey,
      audience,
      authorizedParties,
      signInUrl,
      signUpUrl,
      afterSignInUrl,
      afterSignUpUrl,
      organizationSyncOptions
    } = loadedOptions;
    const requestState = await clerkClient(args, options).authenticateRequest(clerkRequest, {
      apiUrl,
      secretKey,
      jwtKey,
      proxyUrl,
      isSatellite,
      domain,
      publishableKey,
      machineSecretKey,
      audience,
      authorizedParties,
      organizationSyncOptions,
      signInUrl,
      signUpUrl,
      afterSignInUrl,
      afterSignUpUrl,
      acceptsToken: "any"
    });
    const locationHeader = requestState.headers.get(constants.Headers.Location);
    if (locationHeader) {
      handleNetlifyCacheInDevInstance({
        locationHeader,
        requestStateHeaders: requestState.headers,
        publishableKey: requestState.publishableKey
      });
      return new Response(null, { status: 307, headers: requestState.headers });
    }
    if (requestState.status === AuthStatus.Handshake) {
      throw new Error("Clerk: handshake status without redirect");
    }
    args.context.set(authFnContext, (options2) => requestState.toAuth(options2));
    args.context.set(requestStateContext, requestState);
    const response = await next();
    if (requestState.headers) {
      requestState.headers.forEach((value, key) => {
        response.headers.append(key, value);
      });
    }
    return response;
  };
};
export {
  authFnContext,
  clerkMiddleware,
  requestStateContext
};
