// theme-watcher/dist/index.js
// src/ThemeWatcher.tsx
import { useEffect, useRef } from "react";

// src/store.ts
var MEDIA = "(prefers-color-scheme: dark)";
var defaults = {
  storageKey: "theme",
  attribute: "class",
  defaultTheme: "system",
  enableColorScheme: true,
  disableTransitionOnChange: false
};
var cfg = { ...defaults };
var preference = "system";
var resolved = "light";
var system = "light";
var listeners = /* @__PURE__ */ new Set();
var mediaQuery = null;
var mounted = false;
var refCount = 0;
function getSystemTheme() {
  return window.matchMedia(MEDIA).matches ? "dark" : "light";
}
function readStorage() {
  try {
    const val = localStorage.getItem(cfg.storageKey);
    if (val === "light" || val === "dark" || val === "system") return val;
    return null;
  } catch {
    return null;
  }
}
function writeStorage(value) {
  try {
    localStorage.setItem(cfg.storageKey, value);
  } catch {
  }
}
function resolve(pref) {
  return pref === "system" ? system : pref;
}
function applyToDOM(theme) {
  const d = document.documentElement;
  const restore = cfg.disableTransitionOnChange ? disableTransitions() : null;
  if (cfg.attribute === "class") {
    d.classList.remove("light", "dark");
    d.classList.add(theme);
  } else {
    d.setAttribute(cfg.attribute, theme);
  }
  if (cfg.enableColorScheme) {
    d.style.colorScheme = theme;
  }
  restore?.();
}
function disableTransitions() {
  const style = document.createElement("style");
  style.appendChild(
    document.createTextNode(
      "*,*::before,*::after{-webkit-transition:none!important;-moz-transition:none!important;-o-transition:none!important;-ms-transition:none!important;transition:none!important}"
    )
  );
  document.head.appendChild(style);
  return () => {
    window.getComputedStyle(document.body);
    setTimeout(() => {
      document.head.removeChild(style);
    }, 1);
  };
}
function emit() {
  updateSnapshot();
  for (const fn of listeners) fn();
}
function update() {
  resolved = cfg.forcedTheme ?? resolve(preference);
  applyToDOM(resolved);
  emit();
}
function onMediaChange(e) {
  system = e.matches ? "dark" : "light";
  if (preference === "system" && !cfg.forcedTheme) {
    resolved = resolve(preference);
    applyToDOM(resolved);
  }
  emit();
}
function onStorage(e) {
  if (e.key !== cfg.storageKey) return;
  const val = e.newValue;
  if (val === "light" || val === "dark" || val === "system") {
    preference = val;
  } else {
    preference = cfg.defaultTheme;
  }
  update();
}
function attach() {
  if (mounted) return;
  mounted = true;
  system = getSystemTheme();
  mediaQuery = window.matchMedia(MEDIA);
  mediaQuery.addEventListener("change", onMediaChange);
  window.addEventListener("storage", onStorage);
}
function detach() {
  if (!mounted) return;
  mounted = false;
  mediaQuery?.removeEventListener("change", onMediaChange);
  window.removeEventListener("storage", onStorage);
  mediaQuery = null;
}
function mount(overrides) {
  refCount++;
  cfg = { ...defaults, ...overrides };
  if (refCount === 1) {
    attach();
  }
  const stored = readStorage();
  preference = cfg.forcedTheme ? preference : stored ?? cfg.defaultTheme;
  resolved = cfg.forcedTheme ?? resolve(preference);
  applyToDOM(resolved);
  updateSnapshot();
  emit();
  return () => {
    refCount--;
    if (refCount <= 0) {
      refCount = 0;
      detach();
    }
  };
}
function configure(overrides) {
  cfg = { ...defaults, ...overrides };
  const stored = readStorage();
  preference = cfg.forcedTheme ? preference : stored ?? cfg.defaultTheme;
  resolved = cfg.forcedTheme ?? resolve(preference);
  applyToDOM(resolved);
  emit();
}
function setTheme(next) {
  if (next !== "light" && next !== "dark" && next !== "system") return;
  preference = next;
  writeStorage(next);
  update();
}
function toggleMode() {
  const next = resolved === "light" ? "dark" : "light";
  preference = next;
  writeStorage(next);
  update();
}
function getTheme() {
  return readStorage() ?? cfg.defaultTheme;
}
var snapshot = {
  theme: preference,
  resolvedTheme: resolved,
  systemTheme: system
};
function updateSnapshot() {
  snapshot = { theme: preference, resolvedTheme: resolved, systemTheme: system };
}
function getSnapshot() {
  return snapshot;
}
function subscribe(fn) {
  listeners.add(fn);
  return () => {
    listeners.delete(fn);
  };
}

// src/ThemeWatcher.tsx
function ThemeWatcher({
  theme,
  defaultTheme = "system",
  storageKey = "theme",
  attribute = "class",
  enableColorScheme = true,
  disableTransitionOnChange = false
}) {
  const firstMount = useRef(true);
  useEffect(() => {
    return mount({
      defaultTheme,
      storageKey,
      attribute,
      enableColorScheme,
      disableTransitionOnChange,
      forcedTheme: theme
    });
  }, []);
  useEffect(() => {
    if (firstMount.current) {
      firstMount.current = false;
      return;
    }
    configure({
      defaultTheme,
      storageKey,
      attribute,
      enableColorScheme,
      disableTransitionOnChange,
      forcedTheme: theme
    });
  }, [theme, defaultTheme, storageKey, attribute, enableColorScheme, disableTransitionOnChange]);
  return null;
}

// src/useTheme.ts
import { useSyncExternalStore } from "react";
var serverSnapshot = { theme: "system", resolvedTheme: "light", systemTheme: "light" };
function useTheme() {
  const state = useSyncExternalStore(subscribe, getSnapshot, () => serverSnapshot);
  return {
    ...state,
    setTheme,
    set: setTheme,
    get: getTheme,
    toggleMode
  };
}
export {
  ThemeWatcher,
  useTheme
};
