diff --git a/.dist/vite.js b/.dist/vite.js
index cce88a0246b3f7a61af8d8405c0f67ea413b6529..e70c5f672f29bc05309935c35adbc19e11275991 100644
--- a/.dist/vite.js
+++ b/.dist/vite.js
@@ -1,4 +1,5 @@
 import { createRequire } from "node:module";
+import serverless_chromium from "@sparticuz/chromium";
 import { Context, Data, Effect, Layer, Match } from "effect";
 import { createServer } from "vite";
 import { spawn } from "node:child_process";
@@ -43,7 +44,10 @@ const InstallChromium = Effect.async((resume) => {
 	return Effect.sync(() => child.kill());
 });
 const LaunchChromium = Effect.tryPromise({
-	try: () => chromium.launch(),
+	try: async () => chromium.launch(process.env.VERCEL ? {
+		args: serverless_chromium.args,
+		executablePath: await serverless_chromium.executablePath()
+	} : void 0),
 	catch: (cause) => new CaptureError({
 		cause,
 		message: "Could not launch Chromium."
