diff --git a/.dist/vite.js b/.dist/vite.js
index cce88a0246b3f7a61af8d8405c0f67ea413b6529..87dcfb3d09b8021d94df98c578ea8fbf64b3a197 100644
--- a/.dist/vite.js
+++ b/.dist/vite.js
@@ -1,4 +1,5 @@
 import { createRequire } from "node:module";
+import serverless_chromium from "@sparticuz/chromium";
 import { Context, Data, Effect, Layer, Match } from "effect";
 import { createServer } from "vite";
 import { spawn } from "node:child_process";
@@ -43,7 +44,10 @@ const InstallChromium = Effect.async((resume) => {
 	return Effect.sync(() => child.kill());
 });
 const LaunchChromium = Effect.tryPromise({
-	try: () => chromium.launch(),
+	try: async () => chromium.launch(process.env.VERCEL ? {
+		args: serverless_chromium.args,
+		executablePath: await serverless_chromium.executablePath()
+	} : void 0),
 	catch: (cause) => new CaptureError({
 		cause,
 		message: "Could not launch Chromium."
@@ -123,6 +127,10 @@ const AcquireCaptureServer = (config) => Effect.acquireRelease(Effect.tryPromise
 			clearScreen: false,
 			configFile: config.configFile ?? void 0,
 			mode: config.mode,
+			optimizeDeps: {
+				include: [],
+				noDiscovery: true
+			},
 			root: config.root,
 			server: {
 				host: "127.0.0.1",
@@ -157,7 +165,7 @@ const GenerateFromOrigin = (origin, inputs, options) => Effect.gen(function* ()
 			...input,
 			source
 		};
-	}), { concurrency: 4 });
+	}), { concurrency: process.env.VERCEL ? 1 : 4 });
 });
 const GenerationLayers = BrowserCaptureLive.pipe(Layer.merge(ImageEncoderLive));
 const GenerateBuildAssets = (config, inputs, options) => Effect.scoped(Effect.gen(function* () {
